File: gui/menu/compute/compute_scc_decomposition.cpp
#include "compute_scc_decomposition.h"


/* NOTE FOR THE DEVELOPERS:
 * The constructor of a Menu Dialog class deriving from IodeSettingsDialog MUST:
 * 2. call setupUi(this),
 * 3. use the Wrapper classes defined in the wrapper_classes.h header file to wrap the input field items,
 * 4. fill the mapFields member (inherited from IodeSettingsDialog) with the wrapped field items,
 * 5. initialize the inherited className member
 * 6. call the loadSettings() method (inherited from IodeSettingsDialog).
 */


MenuComputeSCCDecomposition::MenuComputeSCCDecomposition(QWidget* parent) : 
    IodeSettingsDialog(parent)
{
    setupUi(this);

	completer = new IodeCompleter(false, false,EQUATIONS, textEdit_equations_list);
	textEdit_equations_list->setCompleter(completer);

    qEquationsList = new WrapperQPlainTextEdit(label_equations_list->text(), *textEdit_equations_list, OPTIONAL_FIELD);
    qTriangulationIterations = new WrapperSpinBox(label_triangulation_iterations->text(), *spinBox_triangulation_iterations, REQUIRED_FIELD);
    qPreRecursiveListName = new WrapperQLineEdit(label_pre_recursive->text(), *lineEdit_pre_recursive_list_name, REQUIRED_FIELD);
    qInterRecursiveListName = new WrapperQLineEdit(label_inter_recursive->text(), *lineEdit_inter_recursive_list_name, REQUIRED_FIELD);
    qPostRecursiveListName = new WrapperQLineEdit(label_post_recursive->text(), *lineEdit_post_recursive_list_name, REQUIRED_FIELD);

    mapFields["EquationsList"] = qEquationsList;
    mapFields["NbIterations"] = qTriangulationIterations;
    mapFields["PreRecursiveListName"] = qPreRecursiveListName;
    mapFields["InterRecursiveListName"] = qInterRecursiveListName;
    mapFields["PostRecursiveListName"] = qPostRecursiveListName;

    className = "MENU_COMPUTE_SCC_DECOMPOSITION";
    loadSettings();
}

MenuComputeSCCDecomposition::~MenuComputeSCCDecomposition()
{
    delete qEquationsList;
    delete qTriangulationIterations;
    delete qPreRecursiveListName;
    delete qInterRecursiveListName;
    delete qPostRecursiveListName;

    delete completer;
}

void MenuComputeSCCDecomposition::compute()
{
    try
    {
        std::string equations_list = qEquationsList->extractAndVerify().toStdString();
        int nb_iterations = qTriangulationIterations->extractAndVerify();
        std::string pre_recursive_list_name = qPreRecursiveListName->extractAndVerify().toStdString();
        std::string inter_recursive_list_name = qInterRecursiveListName->extractAndVerify().toStdString();
        std::string post_recursive_list_name = qPostRecursiveListName->extractAndVerify().toStdString();

        Simulation sim;

        sim.model_calculate_SCC(nb_iterations, pre_recursive_list_name, inter_recursive_list_name, post_recursive_list_name, equations_list);

        this->accept();
    }
    catch (const std::exception& e)
    {
        QMessageBox::critical(this, tr("ERROR"), tr(e.what()));
    }
}

File: gui/iode_gui/menu/compute/compute_scc_decomposition.py
from PySide6.QtCore import Slot
from PySide6.QtWidgets import QWidget, QMessageBox

from iode_gui.settings import MixinSettingsDialog
from iode_gui.text_edit.completer import IodeCompleter
from iode_gui.abstract_main_window import AbstractMainWindow
from .ui_compute_scc_decomposition import Ui_MenuComputeSCCDecomposition

import warnings
from iode import IodeType, Simulation


class MenuComputeSCCDecomposition(MixinSettingsDialog):
    def __init__(self, parent: AbstractMainWindow):
        super().__init__(parent)
        self.ui = Ui_MenuComputeSCCDecomposition()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)  

        self.ui.textEdit_equations_list.setup_completer(iode_types=IodeType.EQUATIONS)

        self.ui.spinBox_triangulation_iterations.setValue(5)
        self.ui.lineEdit_pre_recursive_list_name.setText("_PRE")
        self.ui.lineEdit_inter_recursive_list_name.setText("_INTER")
        self.ui.lineEdit_post_recursive_list_name.setText("_POST")

        self.load_settings()

    @Slot()
    def compute(self):
        try:
            equations_list: str = self.ui.textEdit_equations_list.toPlainText().strip()
            nb_iterations: int = self.ui.spinBox_triangulation_iterations.value()
            pre_recursive_list_name: str = self.ui.lineEdit_pre_recursive_list_name.text().strip()
            inter_recursive_list_name: str = self.ui.lineEdit_inter_recursive_list_name.text().strip()
            post_recursive_list_name: str = self.ui.lineEdit_post_recursive_list_name.text().strip()

            warnings.simplefilter("error")
            simu = Simulation()
            simu.model_calculate_SCC(nb_iterations, pre_recursive_list_name, inter_recursive_list_name, 
                                     post_recursive_list_name, equations_list)
            warnings.simplefilter("default")

            self.accept()
        except Exception as e:
            self.parent().display_output(f"ERROR -> {str(e)}")
            QMessageBox.warning(self, "WARNING", str(e))
            warnings.simplefilter("default")
